src/ASM/BasisFunctionCache.h

```diff
@@ -0,0 +1,151 @@
+
+  //! \brief Return integration scheme nodes.
+  //! \param reduced True to return reduced quadrature
+  const std::array<const double*,Dim>& coord(bool reduced = false) const
+  { return reduced ? reducedQ->xg : mainQ->xg; }
+
+  //! \brief Returns whether or not a reduced quadrature is enabled.
+  bool hasReduced() const { return !reducedQ->gpar[0].empty(); }
+
+  //! \brief Obtain a single integration point parameter.
+  //! \param dir Direction of for integration point
+  //! \param el Element number in given direction
+  //! \param gp Integration point in given direction
+  //! \param reduced True to return parameter for reduced quadrature
+  virtual double getParam(int dir, size_t el, size_t gp, bool reduced = false) const;
+
+protected:
+  //! \brief Template struct holding information about a quadrature.
+  struct Quadrature {
+    std::array<Matrix,Dim> gpar; //!< Array with global integration point parameters
+    std::array<int,Dim> ng; //!< Number of integration point in each dimension
+    std::array<const double*,Dim> wg; //!< Integration scheme weights in each dimension
+    std::array<const double*,Dim> xg;  //!< Integration scheme nodes in each dimension
+
+    //! \brief Clears out configured quadrature.
+    void reset()
+    {
+      for (Matrix& g : gpar)
+        g.clear();
+      ng.fill(0);
+      xg.fill(nullptr);
+      wg.fill(nullptr);
+    }
+  };
+
+  //! \brief Class-specified initialization.
+  virtual bool internalInit() = 0;
+
+  //! \brief Class-specified cleanup.
+  virtual void internalCleanup() = 0;
+
+  //! \brief Calculates basis function info in a single integration point.
+  //! \param el Element of integration point (0-indexed)
+  //! \param gp Integratin point on element (0-indexed)
+  //! \param reduced If true, returns values for reduced integration scheme
+  virtual BasisFunctionVals calculatePt(size_t el, size_t gp,
+                                        bool reduced = false) const = 0;
+
+  //! \brief Calculates basis function info in all integration points.
+  virtual void calculateAll() = 0;
+
+  //! \brief Obtain global integration point index.
+  //! \param el Element of integration point (0-indexed)
+  //! \param gp Integration point on element (0-indexed)
+  //! \param reduced If true return index for reduced integration scheme
+  virtual size_t index (size_t el, size_t gp, bool reduced) const;
+
+  //! \brief Obtain structured integration point indices.
+  //! \param gp Global integration point
+  //! \param reduced If true, returns values for reduced integration scheme
+  std::array<size_t,Dim> gpIndex(size_t gp, bool reduced) const;
+
+  ASM::CachePolicy policy; //!< Cache policy to use
+  std::vector<BasisFunctionVals> values; //!< Cache for main quadrature
+  std::vector<BasisFunctionVals> valuesRed; //!< Cache for reduced quadrature
+  const Integrand* integrand = nullptr; //!< Integrand to use
+  int nderiv = 0; //!< Number of derivatives
```

Comment:
Could see the the `nderiv` member is used for anything useful?

Reply:
https://github.com/OPM/IFEM/pull/571/commits/335388897429e9e482355e7436ceaa58e88eb233#diff-ac311749e8021ce36438b03133af6e7915f698659cbb3e9d354a201cfd075424R41

Reply:
Still don't see it. That link only shows the whole commit containing the added code.

Reply:
hrm. it's used in BasisFunctionCache::init to check if we are requesting more derivatives than what was previously cached. this happens in real application, e.g. when calculating norms.

Reply:
yes, I can see you check it in the init-method. I was confused because the initializations done here do not appear to depend on the actual value on nderiv=nd, the same thing is done nomatter what, unless you return because nd is not larger than last call. But never mind, guess it works as anticipated although the logic in unclear.